Csv

SupriyaGadigone opened the pull request:
The way this works is:
1. Click the export button 
2. In the results.js file, populate and send the data as per needed 
3. Send the data through the url - I realize this isn't the cleanest way to do it but I couldn't think of anything els. Also I looked into it, about 8000 words can be sent through a url request so we should be good for this scope of the project
4. Exporting the file as csv in HomeController with the data received 

The way it looks is:

When values are selected to be exported:
![Screen Shot 2019-03-18 at 3 39 31 PM](https://user-images.githubusercontent.com/15680487/54561485-24f05700-499b-11e9-8b21-486b2071940d.png)

The CSV file:
![Screen Shot 2019-03-18 at 3 39 53 PM](https://user-images.githubusercontent.com/15680487/54561510-30438280-499b-11e9-9b58-070189cd3c9f.png)

When no values are selected to export:
![Screen Shot 2019-03-18 at 3 40 16 PM](https://user-images.githubusercontent.com/15680487/54561527-39ccea80-499b-11e9-8db8-316b5d56bad2.png)

I realize the UI isn't great but just wanted to have something basic working since code reviews might take some time.
 
Closes #16

SupriyaGadigone committed:
Basic CSV works

SupriyaGadigone committed:
New line in CSV file

SupriyaGadigone committed:
Edge cases handled

SupriyaGadigone committed:
Added comment

SupriyaGadigone committed:
Fixed merge conflits

andrewdlqnguyen commented on src/main/resources/templates/fragments/results.html:
```diff
@@ -3,6 +3,13 @@
 <head>
     <meta charset="UTF-8">
     <title>Learning Outcomes Management System</title>
+    <link href="../../static/css/bootstrap.css" rel="stylesheet" th:href="@{/css/bootstrap.css}"/>
```
Dont think you need to add these link to results,html since it's populated into index.html with the link calls already

K-shamG replied:
^I believe that's true as well, @SupriyaGadigone can you test this so that we don't have unnecessary links in results?

andrewdlqnguyen reviewed (approved):
Everything else looks goood :D

K-shamG commented on src/main/resources/static/js/results.js:
```diff
@@ -0,0 +1,50 @@
+    var url = "/export/";
+
+    //get the selected values to export CSV
+    var selectedVals = $("div.card-body input:checked").next('label').map(function () {
```
Might better to give that card-body an ID because otherwise, this select is unclear. Could also cause issues if we had another card-body with inputs.

K-shamG commented on src/main/resources/static/js/results.js:
```diff
@@ -0,0 +1,50 @@
+}
+
+function populateCsvVals(csvExportValsArr, colNum) {
+    $("#courseResults tr").each(function () {
```
I'm not seeing anything with the ID courseResults, the table's ID is results-table. Is it possible this is a mistake, because right now my CSV shows nothing: 

![image](https://user-images.githubusercontent.com/15951317/54608212-9de8c080-4a26-11e9-9fa7-e2aeb6572cdd.png)

K-shamG commented on src/main/resources/templates/fragments/results.html:
```diff
@@ -25,23 +32,25 @@ <h5 class="modal-title" id="exampleModalCenterTitle">Export to CSV</h5>
                                         <form>
                                             <div class="custom-control custom-checkbox">
-                                                <input type="checkbox" class="custom-control-input" id="courseCheck" checked>
+                                                <input type="checkbox" class="custom-control-input" id="courseCheck">
```
I had these as "checked" because I thought a user by default would want all rows exported rather than a default of non of them. Is there a reason you took this out?

K-shamG reviewed (changes requested):
I see the functionality for simple cases is working well (when the table name in results.js is changed to results-table which is something that changed in my pull request for sorting). 

I tried a CSV with a course that has multiple LO's and I don't think this is the ideal output: 
![image](https://user-images.githubusercontent.com/15951317/54608742-c4f3c200-4a27-11e9-9ef7-eaa969f16035.png)
I think we should try to fix that organization on how we are displaying it in the CSV, but for deliverable 2, I would be fine with at least just having a space between LO's if there are multiple in one box. 

However, if a class has multiple categories associated with it, each with their own LOs, the same problem occurs where the organization is confusing. 

Another note is that I feel like it's not intuitive to just have the program say "File has been exported". A user might not know where it's outputted and I'm not sure what the default will be when this is clicked while deployed on heroku. 

When it is outputted, the items show up horizontally which is also a bit strange as I would normally expect the data to go downwards. 

Finally, I'm wondering why it goes through the controller at all. Since all the data is being collected with javascript anyway, could a file not just be created from there instead? I used jquery datatables for the sorting function, looks like it has CSV stuff too. This is a simpler implementation: https://datatables.net/extensions/buttons/examples/initialisation/export.html

SupriyaGadigone committed:
Addressed comments

SupriyaGadigone commented:
Addressed most of the comments except for:
1. I spent couple of hours trying to use jquery datatables but I couldn't figure it out so choose to do it this way.
2. The output for multiple LO's and multiple categories is still not the right way. I don't see any way to fix it the way I have it implemented now and in general people have had issues with this (https://github.com/DataTables/Buttons/issues/60). Adding a space is also going to be hard since for CSV export, all the functionality is based on cell values. I will for sure get back to this once I have fixed things in my other pull request.

![Screen Shot 2019-03-19 at 7 46 05 PM](https://user-images.githubusercontent.com/15680487/54649216-ab7e6480-4a7f-11e9-994c-a64d30095ad8.png)

SupriyaGadigone commented on src/main/resources/static/js/results.js:
```diff
@@ -1,50 +1,96 @@
-    //replaces the specific fragment
-    $("#exportCsv").load(encodeURI(url));
+}
+
```
I found this code (here to below) on stackoverflow and just tweaked it to what we need

K-shamG reviewed (approved):
Looks good for Deliverable 2! Nice work :)
